New target: FrSky/Rotorflight Vantac RF007

functionpointer opened the pull request:
### **User description**
Adds support for [FrSky/Rotorflight Vantac RF007](https://www.rotorflight.org/docs/controllers/frsky-007).

This is a similar flight controller to the [Nexus X](https://github.com/iNavFlight/inav/pull/11047), but it comes with a FrSky receiver instead. While originally designed for helicopters, its form factor is also appealing for fixed wing aircraft running INAV.

Testing
--------

I have tested:

- all 9 servo pins
- "A" port UART
- "C" port UART
- UART1 on AUX/SBUS pins
- internal receiver binds and works using FBUS
- calibrated external voltage input using a multimeter
- verified both gyro and accelerometer work and are oriented correctly
- verified barometer works
- verified Blackbox works
- verified USB-C port 5.1k resistors exist

Choice of pin functions
--------------------------

Many pins can have multiple roles.
I have created three targets `VANTAC_RF007`, `VANTAC_RF007_9SERVOS` and `VANTAC_RF007_NOI2C`.  
`VANTAC_RF007` has 7 servo outputs, 2 UARTs and 1 non-shared I2C.
`VANTAC_RF007_9SERVOS` has 9 servo outputs, 1 UART and non-shared I2C.
`VANTAC_RF007_9SERVOS` has 9 servo outputs, 2 UARTs and no I2C.

I hope this covers most use cases without creating too many targets. I am happy to modify the targets if a different set of tradeoffs is preferred.
The targets are inspired by https://github.com/iNavFlight/inav/pull/11047.


___

### **PR Type**
New Target


___

### **Description**
- Adds support for FrSky/Rotorflight Vantac RF007 flight controller

- Provides three target variants with different pin configurations
  - VANTAC_RF007: 7 servo outputs, 2 UARTs, shared I2C
  - VANTAC_RF007_9SERVOS: 9 servo outputs, 1 UART, I2C
  - VANTAC_RF007_NOI2C: 9 servo outputs, 2 UARTs, no I2C

- Includes built-in FrSky FBUS receiver on UART5

- Supports ICM42688P gyro/accelerometer, SPL06 barometer, W25N01G flash storage


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["Vantac RF007<br/>Flight Controller"] --> B["STM32F722XE"]
  B --> C["ICM42688P<br/>IMU"]
  B --> D["SPL06<br/>Barometer"]
  B --> E["W25N01G<br/>Flash Storage"]
  B --> F["FrSky FBUS<br/>Receiver"]
  B --> G["3 Target<br/>Variants"]
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>target.h</strong><dd><code>Target hardware configuration and pin definitions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/target/VANTAC_RF007/target.h

<ul><li>Defines board identifier F7B5 and USB product string<br> <li> Configures SPI1 for ICM42688P IMU with CW90_DEG alignment<br> <li> Sets up I2C devices with conditional compilation for three variants<br> <li> Defines UART configurations (UART1, UART3, UART4, UART5) with <br>variant-specific pins<br> <li> Configures ADC for battery voltage monitoring on PC2<br> <li> Enables DShot, SerialShot, ESC sensor, and SmartPort master features<br> <li> Sets MAX_PWM_OUTPUT_PORTS to 7 or 9 depending on variant</ul>


</details>


  </td>
  <td><a href="https://github.com/iNavFlight/inav/pull/11082/files#diff-17fd4c3a402f6de39c0e26aa0091068f3bf8ff7309bad2e5d5c9179e32444174">+160/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>target.c</strong><dd><code>Timer hardware and servo output mappings</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/target/VANTAC_RF007/target.c

<ul><li>Defines timer hardware mappings for 7-9 servo outputs<br> <li> Maps S1-S3 to TIM3 channels, TAIL to TIM2, ESC to TIM1<br> <li> Configures RPM and TLM outputs on TIM5 (clash with UART2)<br> <li> Conditionally adds AUX and SBUS outputs on TIM4 for 9-servo variants<br> <li> Uses TIM_USE_OUTPUT_AUTO for flexible servo assignment</ul>


</details>


  </td>
  <td><a href="https://github.com/iNavFlight/inav/pull/11082/files#diff-7b2c905471a301f27a0c09358d4eb3862b349d2b0d50330647ecfd6a4167be62">+45/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>CMakeLists.txt</strong><dd><code>CMake build configuration for variants</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/target/VANTAC_RF007/CMakeLists.txt

<ul><li>Defines build targets for all three Vantac RF007 variants<br> <li> All variants use STM32F722XE microcontroller</ul>


</details>


  </td>
  <td><a href="https://github.com/iNavFlight/inav/pull/11082/files#diff-7000359f811adc842d938da1b761d4c6beb718d099043f986f411d0a092cf82e">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Miscellaneous</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>config.c</strong><dd><code>Target configuration function stub</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/target/VANTAC_RF007/config.c

<ul><li>Provides empty targetConfiguration function for future customization</ul>


</details>


  </td>
  <td><a href="https://github.com/iNavFlight/inav/pull/11082/files#diff-7362a8fec8ddb9dc44416dcab6b3e170733f6de75f956c523678a4b78b7ba11d">+28/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Comprehensive documentation and pin reference</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/target/VANTAC_RF007/README.md

<ul><li>Documents three flight controller variants and their differences<br> <li> Provides pin configuration tables for all three target variants<br> <li> Includes hardware layout with STM32 pin mappings and timer assignments<br> <li> Notes Port C data pin swap discrepancy with manufacturer documentation<br> <li> References Rotorflight documentation and FrSky manual<br> <li> Explains built-in FrSky FBUS receiver on UART5 and bind procedure</ul>


</details>


  </td>
  <td><a href="https://github.com/iNavFlight/inav/pull/11082/files#diff-392979443bf7bee7508c7a1ec42b884f93390d629980527d58dbab964b7b3bcb">+69/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

functionpointer committed:
VANTAC_RF007: Add new target FrSky/Rotorflight Vantac RF007

Quite similar to NEXUSX, but with FrSky built-in receiver

robthomson commented:
Looking at this and pin-out.

We coud achieve:

THR
AIL
ELV
RUDD
PAN
TILT
ROLL

*7 channels all up

4 uarts.

This would essentially be a good combo for many

robthomson commented:
For traditional heli use, the servo ports will always be at the rear of the heli.

robthomson commented:
I will raise it with frsky tomorrow

functionpointer commented:
It seems sharing pins works, but has strange priorities:

1. UART RX
2. I2C
3. UART sensor and peripheral
4. Servo

This means any pins defined as I2C cannot be used for a UART sensor or peripheral like GPS, OSD, FrSky Master or external Blackbox. That is because the pins will be used as I2C instead. There doesn't seem to be a way to disable I2C from the configurator. I found this by testing with an external serial rx, FrSky Master mode + oscilloscope, and using the `resource` command suggested by @sensei-hacker.

On this particular hardware, the priority issue only affects port "C".
Therefore, I have chosen to create a unified target anyway.
It enables 4 of the rows in my table to be reached with a single target.
Specifically, RPM/TLM pins can be Servo or UART2, and AUX/SBUS can be Servo or UART1.
Port "C" is I2C2 by default. It can become UART3, but will currently only do so for a serial rx.

I have updated the [Readme](https://github.com/functionpointer/inav/blob/vantac_rf007/src/main/target/VANTAC_RF007/README.md) to explain how the pin assignment works.

Thanks to @sensei-hacker for pointing out that combining targets like this is possible

functionpointer committed:
VANTAC_RF007: combine targets

functionpointer committed:
VANTAC_RF007: Fix UART3/I2C2 resource conflict on port C

functionpointer committed:
VANTAC_RF007: USE_DSHOT_DMAR, modify defaults

Apparently DSHOT_DMAR can only work with max 4 timers.
With the previous timer setup this caused a limit of 7 motors.
The new setup increases it 9.

However, it costs some flexibility.
The new default config sets up TIM1 as MOTOR by default,
creating sane defaults for platforms with 1 or 4 motors.

functionpointer committed:
VANTAC_RF007: Add MAG support

sensei-hacker commented:
* [ ] Samples received
* [ ] Flash firmware
* [ ] Calibrate
* [ ] Orientation matches
* [ ] Gyro working
* [ ] Accel working
* [ ] Baro working
* [ ] Blackbox
* [ ] UART1
* [ ] UART2
* [ ] UART3
* [ ] UART4
* [ ] UART5
* [ ] UART6
* [ ] LEDs working
* [ ] Buzzer working
* [ ] Motor outputs
* [ ] DShot support on m1-4
* [ ] Servo outputs
* [ ] Voltage ADC
* [ ] Current ADC
* [ ] Analog RSSI
* [ ] Mag I2C Bus
* [ ] Analog Camera working
* [ ] Video Out working
* [ ] OSD working
* [ ] PINIO1
* [ ] PINIO2
